Single epoll thread per 28 cores

adamsitnik opened the pull request:
Edit: this PR has evolved over time, please see the next comments for accurate description

adamsitnik committed:
use struct wrapper for better perf

adamsitnik committed:
check the most common case first

don't access static variable in a loop

adamsitnik committed:
simplify the source code

adamsitnik committed:
use Span instead of raw pointers

adamsitnik committed:
force inline HandleSyncEventsSpeculatively

tmds commented:
kestrel-linux-transport doesn't use ConcurrentDictionary, instead a regular Dictionary with a lock is used. The lookup is performed up-front, which improves locality.

Previous benchmarks for ConcurrentDictionary vs Dictionary+lock showed only small difference. Maybe we'll see a bigger difference for this scenario.

kouvel commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -333,38 +335,32 @@ private void EventLoop()
+                                ev = default;
                             }
+
+                            context = null;
```
Might also need to set `contextWrapper = default`, I suspect that would be a different local. Maybe keep the comment on this section as well.

stephentoub replied:
Is this nulling out really necessary?  We're generally not concerned with nulling out values like this because of debug builds and the like.

kouvel commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:
```diff
@@ -2067,4 +2068,16 @@ public static void OutputTrace(string s)
+    // the goal is to have a dedicated generic instantiation and using:
+    // System.Collections.Concurrent.ConcurrentDictionary`2[System.IntPtr,System.Net.Sockets.SocketAsyncContextWrapper]::TryGetValueInternal(!0,int32,!1&)
+    // instead of:
+    // System.Collections.Concurrent.ConcurrentDictionary`2[System.IntPtr,System.__Canon]::TryGetValueInternal(!0,int32,!1&)
```
Curious that this would perform better. Why is the dedicated generic instantiation better?

stephentoub replied:
Do we ever update the value for an existing key in the dictionary?  If we do, this will make updates more expensive, as they'll be forced to allocate a new node in the CD, whereas with a reference type value, the existing node will be used.

as for why a specific generic instantiation would do better, presumably it's because it's avoiding the generic dictionary lookup, or helping with inlining, or something like that?  Often I see a similar optimization applied as a workaround for removing array covariance checks, but that's on writes, which we shouldn't be doing here frequently.

tmds replied:
> Do we ever update the value for an existing key in the dictionary?

We don't. We use incremental keys for each AsyncContext. When we run out of keys, we start a new SocketEngine.

> as for why a specific generic instantiation would do better, presumably it's because it's avoiding the generic dictionary lookup, or helping with inlining, or something like that? 

I'm curious what it is.

tmds commented:
> kestrel-linux-transport doesn't use ConcurrentDictionary, instead a regular Dictionary with a lock is used. The lookup is performed up-front, which improves locality.

Like [EPollAsyncEngine.EPollThread.cs#L85-L104](https://github.com/tmds/Tmds.LinuxAsync/blob/11233fc8fc80862533e319f4d22855347981aace/src/Tmds.LinuxAsync/EPollAsyncEngine.EPollThread.cs#L85-L104)

kouvel commented:
> I got to the point that the biggest bottleneck is ConcurrentDictionary.TryGetValue and before I try any experiments with ConcurrentDictionary

From the graph it looks like the contention is coming from `EventLoop()`, if I'm reading it right that would be contention on the `ConcurrentQueue` and not the `ConcurrentDictionary`. If it's contention on the `ConcurrentQueue` it might be interesting to try having a non-contending overload/implementation of `ConcurrentQueue.TryDequeue` that differentiates between empty, contention, and successful dequeue (without spin-waiting). In the contention case then try to exit `EventLoop()` prematurely to avoid contention after scheduling another work item to replace it. My hope is that the new work item would not run too soon and it may decrease the parallelization a bit to make faster progress on the queue. Just a thought though, may or may not work.

kouvel commented:
> Previous benchmarks for ConcurrentDictionary vs Dictionary+lock showed only small difference. Maybe we'll see a bigger difference for this scenario.

👍 The lock seems to be rarely taken on other paths could be taken here around the whole inner loop with faster lookups.

kouvel commented:
> From the graph it looks like the contention is coming from EventLoop(), if I'm reading it right that would be contention on the ConcurrentQueue

Ah I didn't read that right, nevermind

kouvel commented:
> I am working on a PR that is going to set the number of epoll threads to 1.

> While testing it for a big number of clients (20k) I've noticed a regression, for JSON Platform benchmark the RPS dropped from 780k to 715k.

Was the 20K clients test also with 1 epoll thread, or would it use 20? I figure with 1 epoll thread `ConcurrentDictionary` accesses shouldn't be contending at all, so there must have been more epoll threads. With 20K connections does it perform better or worse with 1 epoll thread? Lock still may be better in both cases.

adamsitnik commented:
> if I'm reading it right that would be contention on the ConcurrentQueue and not the ConcurrentDictionary

15.64% of time is spent in `ConcurrentDictionary.TryGetValue` while only 1.41% in `ConcurrentQueueSegment.TryEnqueue`

![obraz](https://user-images.githubusercontent.com/6011991/81001582-3ff48380-8e48-11ea-88d9-4ba362946a7e.png)

adamsitnik commented:
> Was the 20K clients test also with 1 epoll thread, or would it use 20? 

For 20k connections from a single load machine:

Before your change it was 740k RPS with 14 epoll threads (Cores / 2)
After your change it was 780k RPS with 14 epoll threads (Cores / 2)

With your change and single epoll thread it dropped to 715k RPS, with the microoptimizations from this PR it's 740k again.

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:
```diff
@@ -1986,7 +1986,8 @@ public SocketError SendFileAsync(SafeFileHandle fileHandle, long offset, long co
         // lead to thread pool starvation issues if the synchronous operations are blocking thread pool threads (typically not
         // advised) and more threads are not immediately available to run work items that would release those operations.
-        public unsafe Interop.Sys.SocketEvents HandleSyncEventsSpeculatively(Interop.Sys.SocketEvents events)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
This alone makes a measurable impact?

adamsitnik replied:
> This alone makes a measurable impact?

Yes, it does.

stephentoub replied:
How much?  I'm not against it (in particular because there's only one call site), but if this actually matters, that suggests the overhead of a single non-virtual, non-generic method call is measurable... and we have a lot of those.  I'm wondering why this one in particular is meaningful.

tmds replied:
This is the only method on the epoll thread that gets called per event. And when there are no sync-over-async operations (like TE benchmarks) it doesn't call other methods.

The improvement may be 'boosted' when the epoll thread is on the edge of being a bottleneck (like with the speculative optimization).

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -333,38 +335,32 @@ private void EventLoop()
+
+                            context = null;
+                        }
+                        else if (handle == shutdownHandle)
```
Do we have anything guaranteeing that shutdownHandle doesn't end up in the dictionary?  Previously that would have been caught as this check was first, but now it won't.  At a minimum we should have an assert earlier.

adamsitnik replied:
> Do we ever update the value for an existing key in the dictionary

The dictionary is updated only when a new socket is accepted. So I would say that reads are much more common than writes

adamsitnik replied:
> Is this nulling out really necessary? We're generally not concerned with nulling out values like this because of debug builds and the like

I wanted to make sure I keep the behaviour the same as in most recent @kouvel PR. I must agree, that I would prefer not to do that and in perfect scenario just avoid all null checks - if `TryGetValue` returns `true` then it's always not null (we control the input values and can ensure that).

adamsitnik replied:
> Previously that would have been caught as this check was first, but now it won't. At a minimum we should have an assert earlier.

Good point. I am going to check this and add an assert or revert this particular change if we can't ensure it.

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -333,38 +335,32 @@ private void EventLoop()
 
                     bool enqueuedEvent = false;
-                    for (int i = 0; i < numEvents; i++)
+                    foreach (var socketEvent in new ReadOnlySpan<Interop.Sys.SocketEvent>(buffer, numEvents))
```
This change was made just for readability / safety?  I'm not seeing how this would have a perf benefit.

adamsitnik replied:
> This change was made just for readability / safety?

Yes, exactly. No perf benefit, just more `Span` and less pointers ;)

adamsitnik committed:
use a Dictionary under a lock instead of ConcurrentDictionary, it improves the perf for ARM and for scenarios with MANY clients

adamsitnik committed:
address code review feedback

adamsitnik committed:
swtich to a single epoll thread by default

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -56,16 +54,21 @@ public bool TryRegister(SafeSocketHandle socket, out Interop.Error error)
+
+        private static int GetEnginesCount()
+        {
+            if (uint.TryParse(Environment.GetEnvironmentVariable("SYSTEM_NET_SOCKETS_ENGINE_COUNT"), out uint count))
```
Can we rename `SYSTEM_NET_SOCKETS_ENGINE_COUNT` to `DOTNET_SYSTEM_NET_SOCKETS_THREAD_COUNT`?  We've been using the DOTNET_ prefix in general, and "thread" is more meaningful than "engine".

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -225,7 +232,14 @@ private void FreeHandle(IntPtr handle)
+
+                lock (_handleToContextMap)
+                {
+                    removed = _handleToContextMap.Remove(handle, out _);
```
Why use this overload rather than just `Remove(handle)`?

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -335,32 +349,38 @@ private void EventLoop()
 
                     bool enqueuedEvent = false;
-                    foreach (var socketEvent in new ReadOnlySpan<Interop.Sys.SocketEvent>(buffer, numEvents))
+                    lock (_handleToContextMap)
```
This isn't a perf issue when there are multiple epoll threads?  That surprises me.

adamsitnik replied:
Each epoll thread has it's own `_handleToContextMap`. It's accessed for a write only when a new socket is allocated or de-allocated (it's rare).  This loop is hot - it's being executed for every epoll notification.

From my initial observations that I am going to confirm with a lot of numbers is that:

* the perf is similar for x64 with not many connections (TechEmpower)
* the perf is better fo x64 with MANY connections
* the perf is better for ARM

stephentoub replied:
> Each epoll thread has it's own _handleToContextMap

Ah. Ok.

> From my initial observations that I am going to confirm with a lot of numbers is that

To my knowledge none of the TE benchmarks open/close sockets at any meaningful frequency.  But we know that happens in some real-world workloads, e.g. `Connection: close` requests.  It'd be good to validate that such use is not negatively impacted.

adamsitnik replied:
> To my knowledge none of the TE benchmarks open/close sockets at any meaningful frequency. But we know that happens in some real-world workloads, e.g. Connection: close requests. It'd be good to validate that such use is not negatively impacted.

@antonfirsov @sebastienros do you have any idea how we could test this scenario?

davidfowl replied:
There's a benchmark for this in our infrastructure cc @sebastienros

sebastienros replied:
We do test that scenario actually by setting a "connection: close" header on the client side. We need to reduce the rate however from the defaults as we reach ports (or sockets) exhaustion quicly by doing that. We might need to find a better methodology if you have an idea for this. We can also build a simple console app that tests that and run it as a client if we know how to do it correctly instead of "connection: close".

sebastienros replied:
We even do this on https to exercise handshakes.

tmds replied:
To reduce the time the lock is held, the lookup could be performed up-front and `SocketAsyncContextWrapper` placed in an array of `EventBufferCount` size.

adamsitnik commented:
Update: from the initial data that I have it looks like switching from concurrent to a regular dictionary under lock (thanks for a great hint @tmds !) combined with the few micro-optimizations is enough to always have a single epoll thread.

I am now going to run the program that is going to run a matrix of:

* Plaintext x Json x Fortunes
* 128, 256, 512, 1024, 20_000 connections
* 12, 28, 32 and 56 core machine
* Before #35330, after #35330, this with #35330

And share the results. If there are no regressions, I am going to ask you for review again. For now please don't merge it.

adamsitnik committed:
code review fixes

kouvel commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -335,32 +343,35 @@ private void EventLoop()
 
                     bool enqueuedEvent = false;
-                    foreach (var socketEvent in new ReadOnlySpan<Interop.Sys.SocketEvent>(buffer, numEvents))
+                    lock (_handleToContextMap)
```
Nit: Use the local instead

kouvel commented:
> 56 core machine

Since that machine has 2 sockets, it might be interesting to try on that machine with `COMPlus_Thread_UseAllCpuGroups=1` and `COMPlus_GCCpuGroup=1`, as those are supported flags for making stuff more numa-aware.

kouvel commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -365,6 +377,10 @@ private void EventLoop()
                                 // such code may keep the stack location live for longer than necessary
                                 context = null;
```
The failure in the test `NonDisposedSocket_SafeHandlesCollected` may be because of the extra reference to the context from the wrapper, might need to also do `contextWrapper = default` to make the test work in debug builds.

stephentoub replied:
That test had been failing since Kount's PR was merged: https://github.com/dotnet/runtime/issues/35846. The JIT will often make tmp copies such that nulling out is often insufficient for debug builds. Hence my wondering why we were doing so.

kouvel replied:
In rolling builds the seems to be failing periodically in release builds, seems like it might be something a bit different but needs to be looked into.

I'm also not too happy about nulling out the locals, seems like that's the only way the test would have passed in debug builds before my change. On one hand the test seemed fair in expecting that a non-disposed socket gets finalized. On another hand that may not happen in debuggable code at least, maybe a better approach would be to make sure it can't happen with quick JIT and enable the test only for release builds or something like that, not sure.

tmds commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:
```diff
@@ -2067,4 +2068,16 @@ public static void OutputTrace(string s)
+    // System.Collections.Dictionary`2[System.IntPtr,System.Net.Sockets.SocketAsyncContextWrapper]::TryGetValueInternal(!0,int32,!1&)
+    // instead of:
+    // System.Collections.Dictionary`2[System.IntPtr,System.__Canon]::TryGetValueInternal(!0,int32,!1&)
+    internal readonly struct SocketAsyncContextWrapper
```
nit: since this is only used in the SocketAsyncEngine as an optimization, maybe the struct can live in that file.

adamsitnik committed:
Revert "use a Dictionary under a lock instead of ConcurrentDictionary, it improves the perf for ARM and for scenarios with MANY clients"

This reverts commit 3dd7784e85d147ca25086bb5a822182aa2c5dbaa.

# Conflicts:
#	src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

adamsitnik committed:
apply code review suggestions

adamsitnik committed:
change the heuristic, single epoll thread is not always enough

adamsitnik commented:
Ok, I have all the data in one place. I'll be editing this post for some time and remove this comment when I am done.

### How to read the results

![obraz](https://user-images.githubusercontent.com/6011991/81195110-b666c200-8fbd-11ea-95ac-e282488478a0.png)

`before #35330` means results before merging #35330
`#35330` means code after merging #35330
`xET yD`  means code after merging #35330 using `x` epoll threads, using `y` Dictionary. `y`:  `C` stands for **C**oncurrent while `L` for generic dictionary used under **L**ock. So `1ET CD` means single epoll thread using Concurrent Dictionary.

`Fortunes Batching` means Fortunes Platform benchmark executed with a copy of `Npgsql.dll` provided by @roji that implements batching

Colors: default MS Excel color scheme where red means the worst and green means the best result.

### x64 12 Cores (the `perf` machine)

Let's start with something simple:

![obraz](https://user-images.githubusercontent.com/6011991/81195394-080f4c80-8fbe-11ea-914c-ef731efb4b1a.png)

As we can see, switching to a single epoll thread and using `ConcurrentDictionary` gives the best results - the `1ET CD` column is the greenest one. No regressions, pure win.

There are two cases where having more epoll threads gives better results:

* JsonPlatform using 512 connections. We could get 130k instead of 128k. The difference is so small that it's ignorable
* PlaintextPlatform using 20_000 connections. The difference is small, but IMHO Plaintext is the most artificial benchmark (because of the pipelining and super small response) and making the heuristic more complex to get few extra % here is not worth it.

adamsitnik commented:
I've shared the numbers from my most recent experiment in a comment above. PTAL

Based on these numbers I came up with the following proposal for the heuristic that determines the number of epoll threads: 

* we need one epoll thread for every 28 cores
* we need to "round up" in a way that 29 cores get 2 epoll threads
* we need to double that for ARM

The code that I've just pushed gives the following results:

ratio is `(#35800/before #35330) - 1.0`


Machine | Connections | Benchmark | before #35330 | #35330 | #35800 | ratio
-- | -- | -- | -- | -- | -- | --
Citrine 28 cores | 128 | PlaintextPlatform | 7,274,914 | 7,389,508 | 7,753,695 | 6.58%
  |   | JsonPlatform | 728,738 | 753,185 | 824,774 | 13.18%
  |   | FortunesPlatform | 288,242 | 293,637 | 301,217 | 4.50%
  |   | Fortunes Batching | 169,087 | 167,237 | 175,886 | 4.02%
  |   |   |   |   |   |  
  | 256 | PlaintextPlatform | 8,855,217 | 8,898,258 | 8,949,456 | 1.06%
  |   | JsonPlatform | 941,176 | 952,582 | 1,078,476 | 14.59%
  |   | FortunesPlatform | 291,339 | 301,213 | 334,134 | 14.69%
  |   | Fortunes Batching | 311,945 | 299,331 | 336,176 | 7.77%
  |   |   |   |   |   |  
  | 512 | PlaintextPlatform | 8,785,644 | 9,139,882 | 9,251,310 | 5.30%
  |   | JsonPlatform | 919,425 | 956,259 | 1,124,823 | 22.34%
  |   | FortunesPlatform | 289,177 | 302,984 | 305,273 | 5.57%
  |   | Fortunes Batching | 358,163 | 349,256 | 411,341 | 14.85%
  |   |   |   |   |   |  
  | 1,024 | PlaintextPlatform | 8,798,429 | 9,093,448 | 9,329,115 | 6.03%
  |   | JsonPlatform | 917,482 | 983,014 | 1,135,564 | 23.77%
  |   | FortunesPlatform | 261,790 | 273,522 | 298,583 | 14.05%
  |   | Fortunes Batching | 372,989 | 374,679 | 407,914 | 9.36%
  |   |   |   |   |   |  
  | 20,000 | PlaintextPlatform | 6,711,039 | 6,707,423 | 7,251,077 | 8.05%
  |   | JsonPlatform | 742,247 | 754,620 | 732,218 | -1.35%
  |   | FortunesPlatform | 208,385 | 220,029 | 227,443 | 9.15%
  |   | Fortunes Batching | 289,530 | 301,026 | 347,141 | 19.90%
  |   |   |   |   |   |  
  |   |   |   |   |   |  
Perf 12 cores | 128 | PlaintextPlatform | 4,548,601 | 4,581,534 | 4,439,358 | -2.40%
  |   | JsonPlatform | 438,914 | 456,929 | 502,222 | 14.42%
  |   | FortunesPlatform | 120,766 | 127,799 | 136,628 | 13.13%
  |   |   |   |   |   |  
  | 256 | PlaintextPlatform | 4,520,799 | 4,728,698 | 5,288,782 | 16.99%
  |   | JsonPlatform | 441,074 | 464,803 | 545,610 | 23.70%
  |   | FortunesPlatform | 123,775 | 132,081 | 138,271 | 11.71%
  |   |   |   |   |   |  
  | 512 | PlaintextPlatform | 4,439,709 | 4,915,243 | 5,368,813 | 20.93%
  |   | JsonPlatform | 456,198 | 480,191 | 554,399 | 21.53%
  |   | FortunesPlatform | 121,289 | 130,383 | 129,009 | 6.36%
  |   |   |   |   |   |  
  | 1,024 | PlaintextPlatform | 4,270,802 | 4,856,757 | 5,282,018 | 23.68%
  |   | JsonPlatform | 453,737 | 480,158 | 561,559 | 23.76%
  |   | FortunesPlatform | 108,143 | 118,506 | 123,213 | 13.94%
  |   |   |   |   |   |  
  | 20,000 | PlaintextPlatform | 3,886,569 | 3,960,775 | 4,039,859 | 3.94%
  |   | JsonPlatform | 309,933 | 333,290 | 388,005 | 25.19%
  |   | FortunesPlatform | 94,303 | 105,309 | 110,534 | 17.21%
  |   |   |   |   |   |  
  |   |   |   |   |   |  
ARM 32 cores | 128 | PlaintextPlatform | 5,325,320 | 5,248,309 | 5,320,333 | -0.09%
  |   | JsonPlatform | 470,719 | 467,996 | 430,931 | -8.45%
  |   | FortunesPlatform | 70,159 | 79,601 | 87,110 | 24.16%
  |   |   |   |   |   |  
  | 256 | PlaintextPlatform | 5,443,043 | 5,433,406 | 5,571,782 | 2.37%
  |   | JsonPlatform | 455,767 | 420,229 | 458,268 | 0.55%
  |   | FortunesPlatform | 73,379 | 76,414 | 86,376 | 17.71%
  |   |   |   |   |   |  
  | 512 | PlaintextPlatform | 5,143,935 | 5,644,389 | 5,773,530 | 12.24%
  |   | JsonPlatform | 425,086 | 397,756 | 454,824 | 7.00%
  |   | FortunesPlatform | 80,027 | 79,361 | 84,626 | 5.75%
  |   |   |   |   |   |  
  | 1,024 | PlaintextPlatform | 5,289,294 | 5,409,985 | 5,817,791 | 9.99%
  |   | JsonPlatform | 350,471 | 376,589 | 440,122 | 25.58%
  |   | FortunesPlatform | 59,300 | 53,292 | 60,414 | 1.88%
  |   |   |   |   |   |  
  | 20,000 | PlaintextPlatform | 3,799,859 | 4,109,911 | 4,229,440 | 11.31%
  |   | JsonPlatform | 246,717 | 258,675 | 309,816 | 25.58%
  |   | FortunesPlatform | 44,415 | 36,242 | 40,573 | -8.65%
  |   |   |   |   |   |  
  |   |   |   |   |   |  
Mono 56 cores | 128 | PlaintextPlatform | 6,011,013 | 6,508,597 | 6,577,628 | 9.43%
  |   | JsonPlatform | 462,300 | 673,968 | 655,741 | 41.84%
  |   |   |   |   |   |  
  | 256 | PlaintextPlatform | 6,896,236 | 6,906,699 | 6,930,573 | 0.50%
  |   | JsonPlatform | 600,973 | 980,908 | 1,052,635 | 75.16%
  |   |   |   |   |   |  
  | 512 | PlaintextPlatform | 6,941,870 | 6,941,820 | 6,954,308 | 0.18%
  |   | JsonPlatform | 623,578 | 1,079,661 | 1,136,029 | 82.18%
  |   |   |   |   |   |  
  | 1,024 | PlaintextPlatform | 6,960,810 | 6,962,596 | 6,957,445 | -0.05%
  |   | JsonPlatform | 741,710 | 1,138,508 | 1,166,838 | 57.32%
  |   |   |   |   |   |  
  | 20,000 | PlaintextPlatform | 6,825,034 | 6,784,191 | 6,786,858 | -0.56%
  |   | JsonPlatform | 660,291 | 919,557 | 944,349 | 43.02%

tmds commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -56,16 +54,34 @@ public bool TryRegister(SafeSocketHandle socket, out Interop.Error error)
+            int result = Environment.ProcessorCount / coresPerSingleEpollThread;
+
+            // and "round" it up, in a way that 29 cores gets 2 epoll threads
+            if (Environment.ProcessorCount % coresPerSingleEpollThread != 0)
```
You could to the rounding up like:
```
int result = ( Environment.ProcessorCount + coresPerSingleEpollThread / 2) / coresPerSingleEpollThread;
```

tmds commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -56,16 +54,34 @@ public bool TryRegister(SafeSocketHandle socket, out Interop.Error error)
+                result *= 2;
+            }
+
+            return Math.Min(result, Environment.ProcessorCount / 2);
```
Won't `result` always be smaller than `Environment.ProcessorCount / 2`?

tmds commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -56,16 +54,34 @@ public bool TryRegister(SafeSocketHandle socket, out Interop.Error error)
+
+            // the data that stands behind this heuristic can be found at https://github.com/dotnet/runtime/pull/35800#issuecomment-624719500
+            // the goal is to have a single epoll thread per every 28 cores
+            const int coresPerSingleEpollThread = 28;
```
The benchmarks confirm an observation we had made from the perftraces. It may interesting to put in the comment:

TechEmpower JSON platform benchmark (which has a low workload per request) shows the epoll thread is fully loaded on a 28-core machine. We add 1 epoll thread per 28 cores to avoid it being a bottleneck.

sebastienros replied:
I can already hear all the complaints about this line ... me being the first concerned. I think we should not have any heuristic with such a value. What if tomorrow we decide to use a different hardware to do benchmarks? We should have some heuristics that are good for the general cases, and allow customer to define custom values that might be better for them. In our case, in the TE repository we would then define an ENV with a number of epoll thread we want. Same for ARM probably, which might depend on each vendor.

tmds replied:
> I can already hear all the complaints about this line ... me being the first concerned.

I guess you refer to the suggestion I made? I prefer to put it explicitly here than to have it implicit in the linked comment.

From the benchmarking we did, TE platform JSON benchmark represents the lowest workload per request. This means the epoll thread will sooner become the bottleneck than on the other benchmarks.

> We should have some heuristics that are good for the general cases

This heuristic is good for the general case, It uses less epoll threads than the previous heuristic (which was a guess with little benchmarking done) and achieves higher performance.

> and allow customer to define custom values that might be better for them.

This is now possible, the count can be set explicitly using the envvar.

sebastienros replied:
The env var is good. But the number `28` within the code is my concern. Unless we say we need to pick one, but 28 because CITRINE should not be the reason IMO.

kouvel replied:
@adamsitnik, on the 56-core machine (2-socket 28-core) are the numbers with or without the env vars `COMPlus_Thread_UseAllCpuGroups=1` and `COMPlus_GCCpuGroup=1`? Without those I suspect it would only be using one CPU group and behaving as a single-socket 28-core machine, with those it should try to use both sockets. The numbers seem to be similar to the 28-core machine and it's a bit odd that 2 epoll threads do better there, though there may be other things going on.

tmds replied:
> but 28 because CITRINE should not be the reason IMO.

We want the value to be high to not have unnecessary epoll threads.
Benchmarks on 28proc x64 (Citrine) showed 1 epoll thread was not a bottleneck.
Benchmarks on 56proc x64 showed 1 epoll thread to be a bottleneck.
So based on these benchmarks: 28 is a safe value, and 56 is not.

stephentoub replied:
I agree with @sebastienros that this value is fairly arbitrary, based on the specific hardware we've tested on.  Does the heuristic hold up on machines with a similar number of cores but a different distribution across nodes?  What about when hyperthreading is disabled?  Did we try it with cloud VMs?

If we believe this is strictly better than the heuristic we currently use, ok.  It just feels a lot like "well it works on my machine" :)

adamsitnik replied:
> on the 56-core machine (2-socket 28-core) are the numbers with or without the env vars

I am always using the default settings (no env vars)

kouvel replied:
It might be common to set those env vars on multi-numa-node machines if the intention is to scale up. Might also be interesting to try the AMD machine with those env vars since it also has multiple numa nodes. Not suggesting for this change or anything but it might provide more insights on heuristics for number of epoll threads.

kouvel replied:
Ahh nevermind, from a brief look it almost looks like all of the the CPU group stuff is disabled on Linux and those env vars may not have any effect.

adamsitnik replied:
> It might be common to set those env vars on multi-numa-node machines if the intention is to scale up. 

Thanks for pointing this out. I am going to run this config as well.

kouvel replied:
Pls see my latest comment :) there might be more work to do there in the VM, I'm not up-to-date on what's happening there

tmds replied:
> I agree with @sebastienros that this value is fairly arbitrary, based on the specific (and limited) hardware we've tested on. Does the heuristic hold up on machines with a similar number of cores but a different distribution across nodes? What about when hyperthreading is disabled? Did we try it with cloud VMs?

We see 1 epoll thread is enough to load a 28 core machine (Citrine) with a benchmark that has low threadpool workload vs epoll workload (TE JSON platform).

That's what is captured by `coresPerSingleEpollThread = 28`.

This heuristic also works well in the likely cases that
- ProcessorCount is lower than 28
- for higher workloads per epoll workload

This heuristic isn't tuned for multi-node machines, or machines with 28+++ procs.

tmds commented:
@stephentoub @adamsitnik I think we can remove the `MinHandlesForAdditionalEngine` logic. It was there to avoid creating too many epoll thread on limited machines. Now we have low nr of epoll threads on such machines anyway.

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -56,16 +54,34 @@ public bool TryRegister(SafeSocketHandle socket, out Interop.Error error)
+
+            // the data that stands behind this heuristic can be found at https://github.com/dotnet/runtime/pull/35800#issuecomment-624719500
+            // the goal is to have a single epoll thread per every 28 cores
+            const int coresPerSingleEpollThread = 28;
```
Nit:
```suggestion
            const int CoresPerThread = 28;
```
- Cores => cores
- We use epoll on Linux but kqueues on mac
- Not sure what "single" meant :)

tmds replied:
`Cores` -> `Processors`

adamsitnik committed:
simplify the heuristic and add a comment

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -56,16 +54,34 @@ public bool TryRegister(SafeSocketHandle socket, out Interop.Error error)
-#pragma warning restore CA1802
+        private static readonly int s_engineCount = GetEnginesCount();
+
+        private static int GetEnginesCount()
```
Nit: GetEngineCount, just to match with s_engineCount

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -56,16 +54,34 @@ public bool TryRegister(SafeSocketHandle socket, out Interop.Error error)
-            Environment.ProcessorCount >= 6 ? Environment.ProcessorCount / 2 : 1;
-#endif
-#pragma warning restore CA1802
+        private static readonly int s_engineCount = GetEnginesCount();
```
Nit: should this be s_maxEngineCount?  We won't always have this many, but we may grow to this many based on the number of concurrent sockets, right?

tmds replied:
@stephentoub I've suggested to remove that logic as part of this PR (https://github.com/dotnet/runtime/pull/35800#issuecomment-624861323).

adamsitnik replied:
@tmds You are most probably right. The only use case for keeping it is a machine with many cores and very few connections. Which should be uncommon.

Would you prefer me to remove it now or would you like to do this in your upcoming PR that is going to enable the "inlining"?

tmds replied:
@adamsitnik remove it here, it is unrelated to inlining.

adamsitnik replied:
@tmds I am going to merge it as it is right now as I would really love to see the update numbers. I am going to send a PR with MinHandles logic removal today or tomorrow.

sebastienros commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -58,29 +58,29 @@ public bool TryRegister(SafeSocketHandle socket, out Interop.Error error)
+                : 30;
 
-            return Math.Min(result, Environment.ProcessorCount / 2);
+            return Math.Max(1, (int)Math.Round(Environment.ProcessorCount / (double)coresPerEngine));
```
> and "round" it up, in a way that 29 cores gets 2 epoll threads

So now anyting below 44 cores on x64 will get 1 thread? Then 2 after 76 cores ...

adamsitnik replied:
Yes and this should be enough for vast majorty of real-life scenarios.

adamsitnik commented:
TechEmpower is super artificial (super small socket reads and writes & extremely high load) and even under such high load, one engine (producer) is capable of keeping busy up to 30 CPU Cores (8 on ARM). This is possible thanks to the amazing work that @kouvel has done in #35330

In real-life scenario, nobody should ever need more than one epoll thread for the entire app. But we can't predict all possible usages and I believe that these numbers (30 & 8) are safe because it would be super hard to generate more network load.

I've simplified the heuristic and added explanation.

adamsitnik committed:
apply the naming suggestions

adamsitnik committed:
Merge remote-tracking branch 'upstream/master' into epollThread

tmds commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -106,7 +122,7 @@ public bool TryRegister(SafeSocketHandle socket, out Interop.Error error)
         private static readonly IntPtr MaxHandles = IntPtr.Size == 4 ? (IntPtr)int.MaxValue : (IntPtr)long.MaxValue;
 #endif
-        private static readonly IntPtr MinHandlesForAdditionalEngine = s_engineCount == 1 ? MaxHandles : (IntPtr)32;
+        private static readonly IntPtr MinHandlesForAdditionalEngine = s_maxEngineCount == 1 ? MaxHandles : (IntPtr)32;
```
We could remove this.

And if it isn't removed, do we want to set this to `1` when the `DOTNET_SYSTEM_NET_SOCKETS_THREAD_COUNT` is set?

tmds replied:
Adam, you can ignore this. I had missed your reply https://github.com/dotnet/runtime/pull/35800#discussion_r421576408.
